File: section-03-basic-maths/14-divisors-of-a-number/divisors-of-a-number-opt.cs
using System;
using System.Collections.Generic;

class Solution
{
    /* Function to find all 
    divisors of n */
    public List<int> divisors(int n)
    {

        // To store the divisors;
        List<int> ans = new List<int>();

        int sqrtN = (int)Math.Sqrt(n);

        // Or simply we could do:-   for (int i = 1; i * n <= n; i++) {

        // Iterate from 1 to sqrtN
        for (int i = 1; i <= sqrtN; i++)
        {

            // If a divisor is found
            if (n % i == 0)
            {
                // Add it to the answer
                ans.Add(i);

                /* Add the counterpart divisor
                 if it's different from i */
                if (i != n / i)
                {
                    ans.Add(n / i);
                }
            }
        }

        // Sorting the result 
        ans.Sort();

        // Return the result
        return ans;
    }
}



/* 

------Time Complexity------

Loop runs till:

√n

So:

Time Complexity = O(√n)

Sorting:

O(k log k)

(where k = number of divisors)

Final:
O(√n + k log k)

------Space Complexity------

We store divisors:

Space Complexity = O(k)

(where k = number of divisors)

 */

File: section-03-basic-maths/14-divisors-of-a-number/divisors-of-a-number-opt.py
import math


class Solution:
    def divisors(self, n):
        ans = []
        for i in range(1, int(math.sqrt(n)) + 1):
            if n % i == 0:
                ans.append(i)

                if i != n // i:
                    ans.append(n // i)

        ans.sort()

        return ans





""" 
Time Complexity
O(sqrt(n) + d log d) where d is the number of divisors of n, because the loop runs sqrt(n) times and sorting the divisors takes d log d
Space Complexity
O(d) where d is the number of divisors of n, as we store all divisors in the ans list


"We iterate only up to √n to find divisor pairs, which takes O(√n) time. For each divisor, we store it in a list, which takes O(d), where d is the number of divisors. Finally, we sort the list of divisors, which takes O(d log d). So overall complexity is O(√n + d log d)."


Sorting takes O(d log d) because Python uses Timsort, which is a comparison-based sorting algorithm. In comparison-based sorting, we need about log d levels of comparisons, and at each level we compare up to d elements. So overall, the number of operations becomes proportional to d log d.

"""